app/controllers/application_controller.rb

```diff
@@ -5,6 +5,121 @@ class ApplicationController < Sinatra::Base
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    enable :sessions
+    set :session_secret, "secret"
   end
 
-end
\ No newline at end of file
+  get '/' do
+    erb :'homepage'
+  end
+
+  get '/signup'  do
+    if !current_user
+      erb :'signup'
+    else
+      redirect to '/tweets'
+    end
+  end
+
+  post '/signup' do
+    if params[:username].empty? || params[:email].empty? || params[:password].empty?
+      redirect to '/signup'
+    else
+      @user = User.create(params)
+      session[:id] = @user.id
+      redirect to '/tweets'
+    end
+  end
+
+  get '/tweets' do
+    if current_user
+      erb :'tweets'
+    else
+      redirect to '/login'
+    end
+  end
+
+  get '/login' do
+    if !current_user
+      erb :'login'
+    else
+      redirect to "/tweets"
+    end
+  end
+
+  post '/login' do
+    user = User.find_by(username: params[:username])
+    if user && user.authenticate(params[:password])
+
+      session[:id] = user.id
+      redirect to '/tweets'
+    else
+      redirect to "/tweets"
+    end
+  end
+
+  get '/logout' do
+    session.clear if current_user
+    redirect to '/login'
+  end
+
+  get '/tweets/new' do
+    if current_user
+      erb :'/tweets/new'
+    else
+      redirect to '/login'
+    end
+  end
+
+  post '/tweets/new' do
```

Comment:
Typically, tweet creation would be posted to `/tweets` instead of `/tweets/new`.